### ISSUE TITLE
isLoading indicator on all the hooks on the auth client

### ISSUE THREAD

**Issue by firstaxel** [better-auth/better-auth#180]
i am making a request to add a loading indicator on the signUp and signIn hooks cause it helps the user experience when loading indicator is shown when any functions is run on the client

**Comment by Loa212** [comment]
Since those are just promises (not hooks), tanstack-query's useMutation can achieve the same result. 

No need to add it here IMHO.

Here's how I would do it:

```tsx
import { useMutation } from '@tanstack/react-query';
import { signIn } from '@/lib/auth-client';

const useSignIn = () => {
  return useMutation({
    mutationFn: async (credentials: { email: string; password: string }) => {
      return await signIn(credentials);
    },
  });
};

// Usage in a component
const SignInComponent: React.FC = () => {
  const { mutate: signIn, isLoading, isError, error } = useSignIn();

  const handleSignIn = () => {
    signIn({ email: 'user@example.com', password: 'password' });
  };

  return (
    <button onClick={handleSignIn} disabled={isLoading}>
      {isLoading ? 'Signing In...' : 'Sign In'}
    </button>
  );
};
```


TLDR: use tanstack

**Comment by Bekacru** [comment]
you can do what @Loa212  recommended here or you can also have your own loading state and set it on a callback.


```ts
const [loading, setLoading] = useState(false)

await signIn.email(
     {
         email,
         password
     },
     {
          onResponse: () => {
                setLoading(false);
           },
          onRequest: () => {
                 setLoading(true);
          },
          onError: (ctx) => {
                toast.error(ctx.error.message);
          },
      }
 )
```

**Comment by firstaxel** [comment]
I tried the tanstack approach but there is a issue associated with that ,
which is you can't do the same with socials cause it need callbackurl which
will not be redirected with tansack query and also with email, there is a
issue with when signing up with email it is complicated . the way I will
use Bereket way but having to tackle with usestate, I just wish we could
have the same time done like tansack query which we have the mutate,
isloading, data, error it will be a good developer experience and also for
the user

**Comment by Loa212** [comment]
You can do it like this:

```tsx
const {data, isPending, error} = useMutation({
    mutationFn: async () => {
         return await client.signIn.social({
           provider: "github",
           callbackURL: "/dashboard",
          })
       },
  });
```

**Comment by Bekacru** [comment]
> I just wish we could have the same time done like tansack query which we have the mutate, isloading, data, error it will be a good developer experience and also for the user

This would require us to turn all the client methods into hooks, which we're not planning to do because we want the client to remain agnostic and consistent across different frameworks.
